graph.go

```diff
@@ -4,193 +4,166 @@
-// be returned as Edge{tail,head}. Thus there's a conflict that cannot be resolved between the
-// two interface requirements.
-type MutableGraph interface {
-	CostGraph
-	Mutable
+	// SetEdge adds an edge from one node to another.
+	// If the nodes do not exist, they are added.
+	// SetEdge will panic if the IDs of the e.From
+	// and e.To are equal.
+	SetEdge(e Edge, cost float64)
 
-	// Like EdgeBetween in Graph, AddUndirectedEdge adds an edge between two nodes.
-	// If one or both nodes do not exist, the graph is expected to add them. However,
-	// if the nodes already exist it should NOT replace existing nodes with e.From() or
-	// e.To(). Overwriting nodes should explicitly be done with another call to AddNode()
-	AddUndirectedEdge(e Edge, cost float64)
+	// RemoveEdge removes the given edge, leaving the
+	// terminal nodes. If the edge does not exist it
+	// is a no-op.
+	RemoveEdge(Edge)
+}
 
-	// RemoveEdge clears the stored edge between two nodes. Calling this will never
-	// remove a node. If the edge does not exist this is a no-op, not an error.
-	RemoveUndirectedEdge(Edge)
+// MutableUndirected is an undirected graph that can be arbitrarily altered.
+type MutableUndirected interface {
+	Undirected
+	Mutable
 }
 
-// MutableDirectedGraph is an interface that ensures one can construct an arbitrary directed
-// graph. Naturally, a MutableDirectedGraph works for both undirected and directed cases,
-// but simply using a MutableGraph may be cleaner. As the documentation for MutableGraph
-// notes, however, a graph cannot safely implement MutableGraph and MutableDirectedGraph
-// at the same time, because of the functionality of a EdgeTo in DirectedGraph.
-type MutableDirectedGraph interface {
-	CostDirectedGraph
+// MutableDirected is a directed graph that can be arbitrarily altered.
+type MutableDirected interface {
+	Directed
 	Mutable
+}
 
-	// Like EdgeTo in DirectedGraph, AddDirectedEdge adds an edge FROM From TO To.
-	// If one or both nodes do not exist, the graph is expected to add them. However,
-	// if the nodes already exist it should NOT replace existing nodes with e.From() or
-	// e.To(). Overwriting nodes should explicitly be done with another call to AddNode()
-	AddDirectedEdge(e Edge, cost float64)
+// WeightFunc is a mapping between an edge and an edge weight.
+type WeightFunc func(Edge) float64
 
-	// Removes an edge FROM e.From TO e.To. If no such edge exists, this is a no-op,
-	// not an error.
-	RemoveDirectedEdge(Edge)
+// UniformCost is a WeightFunc that returns an edge cost of 1 for a non-nil Edge
+// and Inf for a nil Edge.
+func UniformCost(e Edge) float64 {
+	if e == nil {
+		return math.Inf(1)
+	}
+	return 1
 }
 
-// A function that returns the cost of following an edge
-type CostFunc func(Edge) float64
+// CopyUndirected copies nodes and edges as undirected edges from the source to the
+// destination without first clearing the destination. If the source does not
```

Comment:
Could you say "If the source is not a Weighter, then UniformCost is used in the call to SetEdge". I was confused for a bit about how the pieces all interact